### ISSUE TITLE
Enhancement: Add Early Stopping and Learning Rate Scheduling to WhichmutTrainer

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/netam#160]
# Enhancement: Add Early Stopping and Learning Rate Scheduling to WhichmutTrainer

## Summary

The current `WhichmutTrainer` implementation lacks early stopping and learning rate scheduling functionality that is available in the main `framework.py` training infrastructure. This enhancement would add these essential training features to improve model training efficiency and prevent overfitting.

## Background

The existing `netam.framework.Burrito` class (lines 617-618, 766-769, 780, 784-786, 801-802) includes:
- **Early stopping**: Stops training when validation loss doesn't improve for a specified number of epochs
- **Learning rate scheduling**: Uses `ReduceLROnPlateau` to reduce learning rate when validation loss plateaus
- **Best model preservation**: Saves and restores the best model state during training
- **Minimum learning rate check**: Stops training when learning rate falls below threshold

The `WhichmutTrainer` currently only supports basic epoch-by-epoch training without these advanced features.

## Motivation

1. **Prevent Overfitting**: Early stopping based on validation loss helps prevent models from overfitting to training data
2. **Training Efficiency**: Automatic stopping when no improvement is detected saves computational resources
3. **Learning Rate Optimization**: Adaptive learning rate scheduling can improve convergence
4. **Consistency**: Aligning WhichmutTrainer with existing framework patterns ensures consistent behavior across the codebase

### 1. Enhanced WhichmutTrainer Class

Add parameters to `__init__`:
```python
def __init__(self, model, optimizer=None, scheduler=None, 
             min_learning_rate=1e-4, early_stop_patience=20, min_delta=1e-4):
    # Early stopping state tracking
    self.best_val_loss = float('inf')
    self.best_model_state = None
    self.patience_counter = 0
    self.early_stopped = False
```

### 2. New Training Method

Add `train_with_validation()` method:
```python
def train_with_validation(self, train_dataloader, val_dataloader, 
                        epochs, save_best=True, verbose=True):
    """Train with validation, LR scheduling, and early stopping."""
    # Implementation follows framework.py patterns
```

### 3. Enhanced Factory Function

Update `create_whichmut_trainer()` to accept scheduling parameters:
```python
def create_whichmut_trainer(
    model_config_yaml: str,
    device: torch.device,
    # ... existing params ...
    # Scheduler parameters
    scheduler_factor: Optional[float] = None,
    scheduler_patience: Optional[int] = None,
    # Early stopping parameters
    early_stop_patience: Optional[int] = None,
    min_delta: Optional[float] = None,
):
```

### 4. Configuration Integration

Support configuration via YAML with defaults:
```yaml
# Training configuration
learning_rate: 0.001
min_learning_rate: 1e-5
optimizer_name: "Adam"

# Scheduler parameters
scheduler_factor: 0.5      # Reduce LR by 50% when plateau
scheduler_patience: 10     # Wait 10 epochs before reducing LR

# Early stopping parameters  
early_stop_patience: 20    # Stop if no improvement for 20 epochs
min_delta: 1e-4           # Minimum improvement threshold
```

### Key Features

1. **Early Stopping Logic** (following framework.py:784-786):
   - Track best validation loss
   - Increment patience counter when no improvement
   - Stop training when patience exceeded

2. **Learning Rate Scheduling** (following framework.py:617-618, 780):
   - Use `ReduceLROnPlateau` scheduler
   - Step scheduler with validation loss
   - Check for minimum learning rate

3. **Best Model Restoration** (following framework.py:801-802):
   - Save model state when validation improves
   - Restore best model at end of training

4. **Progress Monitoring**:
   - Comprehensive logging of training progress
   - Learning rate, loss, and patience tracking

### Error Handling

- Graceful handling of edge cases (no validation data, empty batches)
- Robust gradient checking and retry logic (following existing patterns)
- Clear error messages and logging

### Backward Compatibility

- All new parameters have sensible defaults
- Existing training workflows continue to work unchanged
- Optional scheduler and early stopping (can be disabled)

## Benefits

1. **Improved Training Quality**: Early stopping prevents overfitting and finds optimal stopping points
2. **Resource Efficiency**: Automatic stopping saves computation time and resources
3. **Better Convergence**: Adaptive learning rate scheduling improves model convergence
4. **Code Consistency**: Aligns WhichmutTrainer with established framework patterns
5. **Flexibility**: Configurable parameters allow fine-tuning for different use cases

## Testing Strategy

1. **Unit Tests**: Test early stopping logic, scheduler integration, and state management
2. **Integration Tests**: Verify compatibility with existing WhichmutCodonDataset and loss functions
3. **Regression Tests**: Ensure existing functionality remains unchanged
4. **Performance Tests**: Validate training efficiency improvements

### Basic Early Stopping
```python
trainer = create_whichmut_trainer(
    config_yaml, device,
    early_stop_patience=15,
    min_delta=1e-4
)
```

### Full Configuration
```python
trainer = create_whichmut_trainer(
    config_yaml, device,
    learning_rate=0.001,
    min_learning_rate=1e-5,
    scheduler_factor=0.5,
    scheduler_patience=10,
    early_stop_patience=20,
    min_delta=1e-4
)

train_losses, val_losses = trainer.train_with_validation(
    train_loader, val_loader, 
    epochs=100, save_best=True
)
```

## Alternative Approaches Considered

1. **Separate EarlyStopping Class**: More modular but adds complexity
2. **Callback System**: More flexible but requires larger architectural changes
3. **Integration with Existing Burrito**: Would require significant refactoring of WhichMut-specific logic

The proposed approach balances functionality, simplicity, and consistency with existing patterns.

## Related Issues

- Consistency with `netam.framework.Burrito` training patterns
- WhichMut training optimization and efficiency
- Model validation and overfitting prevention

## Implementation Priority

**High** - This enhancement would significantly improve WhichMut training quality and efficiency while maintaining consistency with established framework patterns.

**Comment by matsen** [comment]
I like this idea!

I would like it even more if we could take this opportunity to factor things up into an abstract class. We should absolutely have a WhichmutTrainer class which is specialized to the whichmut_loss. But I think we should have an AbstractTrainer class that holds things like the scheduler etc.

I see

> Comprehensive logging of training progress

which is a kettle of fish. Are you leaning towards CSV or Tensorboard?

I also would want to make sure that any backwards compatibility considerations were minimal-- we want the best thing moving forward.
